### PR TITLE
feat(connector): [Iatapay] add upi qr support

### PR DESCRIPTION
**Pull request by AkshayaFoiger**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR introduces a new payment_method_type: `upi_intent` and a new  payment_method_data:  `upi_qr: {}`  along with iatapay qr code implementation.

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

_Note:Test through stripe compatibility layer_
1. Create a upi qr payment 
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_ZLiIlOCsh5GUTiRJCCSc6lOIXpo0tTGuROyxvZokpKSIl5Lh5mCYwBB9IZV3Jno2' \
--data-raw '{
    "amount": 5000,
    "currency": "INR",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 5000,
    "customer_id": "IatapayCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "upi",
    "payment_method_type": "upi_intent",
    "payment_method_data": {
        "upi": {
            "upi_intent": {
                
            }
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "IN",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```
Response

```
{
    "payment_id": "pay_ZBLXkfFi0wcdBPrQkKPL",
    "merchant_id": "merchant_1716310637",
    "status": "requires_customer_action",
    "amount": 5000,
    "net_amount": 5000,
    "amount_capturable": 5000,
    "amount_received": null,
    "connector": "iatapay",
    "client_secret": "pay_ZBLXkfFi0wcdBPrQkKPL_secret_YYAnlcRyF7ZX3r82WtjL",
    "created": "2024-05-22T10:04:05.468Z",
    "currency": "INR",
    "customer_id": "IatapayCustomer",
    "customer": {
        "id": "IatapayCustomer",
        "name": "John Doe",
        "email": "guest@example.com",
        "phone": "999999999",
        "phone_country_code": "+1"
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "upi",
    "payment_method_data": {
        "upi": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "IN",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
      "next_action": {
        "type": "fetch_qr_code_information",
        "qr_code_fetch_url": "https://sandbox.iata-pay.iata.org/api/v1/payments/PJHE6F45O1D7M/checkout/api/qr"
    },
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "upi_intent",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "IatapayCustomer",
        "created_at": 1716372245,
        "expires": 1716375845,
        "secret": "epk_41f95816c9844c2e9c8672ec5bc91d65"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": "P1D74RYCPET6J",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pay_ZBLXkfFi0wcdBPrQkKPL_1",
    "payment_link": null,
    "profile_id": "pro_aqLkyMdeFcYrp3v3yPd8",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_nLDpyk2OF9mrlivQTJnS",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-05-22T10:19:05.468Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-05-22T10:04:08.642Z",
    "frm_metadata": null
}
```

2. Create a upi payment
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key:""' \
--data-raw '{
    "amount": 5000,
    "currency": "INR",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 5000,
    "customer_id": "IatapayCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "upi",
    "payment_method_type": "upi_collect",
    "payment_method_data": {
        "upi": {
            "upi_collect": {
                  "vpa_id": "successtest@ita"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "IN",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}'
```
Response
```
{
    "payment_id": "pay_VfMfCw7jxtbHNc5xvF4Z",
    "merchant_id": "merchant_1716310637",
    "status": "requires_customer_action",
    "amount": 5000,
    "net_amount": 5000,
    "amount_capturable": 5000,
    "amount_received": null,
    "connector": "iatapay",
    "client_secret": "pay_VfMfCw7jxtbHNc5xvF4Z_secret_lMsdWF0m50OBBvxe1fl7",
    "created": "2024-05-22T09:54:19.813Z",
    "currency": "INR",
    "customer_id": "IatapayCustomer",
    "customer": {
        "id": "IatapayCustomer",
        "name": "John Doe",
        "email": "guest@example.com",
        "phone": "999999999",
        "phone_country_code": "+1"
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "upi",
    "payment_method_data": {
        "upi": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "IN",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": {
        "type": "redirect_to_url",
        "redirect_to_url": "http://localhost:8080/payments/redirect/pay_VfMfCw7jxtbHNc5xvF4Z/merchant_1716310637/pay_VfMfCw7jxtbHNc5xvF4Z_1"
    },
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "upi_collect",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "IatapayCustomer",
        "created_at": 1716371659,
        "expires": 1716375259,
        "secret": "epk_7bb67ee9c3304c8a9458ba6ef2488225"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": "P46Y265SXCTXZ",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pay_VfMfCw7jxtbHNc5xvF4Z_1",
    "payment_link": null,
    "profile_id": "pro_aqLkyMdeFcYrp3v3yPd8",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_nLDpyk2OF9mrlivQTJnS",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-05-22T10:09:19.813Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-05-22T09:54:21.635Z",
    "frm_metadata": null
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/connector/iatapay/transformers.rs [hunk 1]
```diff
@@ -291,8 +309,46 @@ fn get_iatpay_response(
     };
     let connector_response_reference_id = response.merchant_payment_id.or(response.iata_payment_id);
 
-    let payment_response_data = response.checkout_methods.map_or(
-        types::PaymentsResponseData::TransactionResponse {
+    let payment_response_data = match response.checkout_methods {
+        Some(checkout_methods) => {
+            let (connector_metadata, redirection_data) =
+                match checkout_methods.redirect.redirect_url.ends_with("qr") {
```
Comment (SamraatBansal) [inline crates/router/src/connector/iatapay/transformers.rs:315]: Is this a reliable check?

**Review by SamraatBansal — APPROVED** [review]
LGTM!

File: crates/router/src/connector/iatapay/transformers.rs (hunk 1 above)
Comment (AkshayaFoiger) [reply crates/router/src/connector/iatapay/transformers.rs:315]: Yes. In all environments, response Url ends with their respective checkout method

**Review by Chethan-rao — DISMISSED** [review]
Routing specific changes LGTM

File: crates/router/src/core/payments/transformers.rs [hunk 2]
```diff
@@ -876,6 +884,18 @@ pub fn qr_code_next_steps_check(
...
 
+pub fn fetch_qr_code_url_next_steps_check(
+    payment_attempt: storage::PaymentAttempt,
+) -> RouterResult<Option<api_models::payments::FetchQrCodeInformation>> {
+    let qr_code_steps: Option<Result<api_models::payments::FetchQrCodeInformation, _>> =
+        payment_attempt
+            .connector_metadata
+            .map(|metadata| metadata.parse_value("FetchQrCodeInformation"));
+
+    let qr_code_fetch_url = qr_code_steps.transpose().ok().flatten();
```
Comment (sahkal) [inline crates/router/src/core/payments/transformers.rs:895]: why are we ignoring the error in case of parse failures?

Comment (AkshayaFoiger) [reply crates/router/src/core/payments/transformers.rs:895]: This function is used to decide the next action,  if `fetch_qr_code_url_next_steps_check` parses connector_metadata to FetchQrCodeInformation. We mark next action into `fetch_qr_code_information`. If it doesn't parse, that implies next action is not `fetch_qr_code_information`

**Review by sahkal — APPROVED** [review]

File: crates/openapi/src/openapi.rs [hunk 3]
```diff
@@ -308,6 +310,7 @@ Never share your secret api keys. Keep them guarded and secure.
         api_models::payments::ApplepayConnectorMetadataRequest,
         api_models::payments::SessionTokenInfo,
         api_models::payments::SwishQrData,
+        api_models::payments::UpiData,
```
Comment (SamraatBansal) [inline crates/openapi/src/openapi.rs:313]: UpiData is already there

**Review by SamraatBansal — APPROVED** [review]

File: crates/openapi/src/openapi.rs (hunk 3 above)
Comment (SamraatBansal) [inline crates/openapi/src/openapi.rs:313]:
```suggestion
```

**Review by SamraatBansal — APPROVED** [review]

**Review by sahkal — APPROVED** [review]

**Review by Chethan-rao — DISMISSED** [review]

File: crates/hyperswitch_domain_models/src/payment_method_data.rs [hunk 4]
```diff
@@ -289,10 +289,20 @@ pub struct CryptoData {
...
 #[serde(rename_all = "snake_case")]
-pub struct UpiData {
+pub enum UpiData {
+    UpiCollect(UpiCollectData),
+    UpiIntent(UpiIntentData),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, serde::Deserialize, serde::Serialize)]
+#[serde(rename_all = "snake_case")]
+pub struct UpiCollectData {
```
Comment (Narayanbhat166) [inline crates/hyperswitch_domain_models/src/payment_method_data.rs:299]: we have changed a struct to enum, is this change backwards compable? If this breaks existing api contract then this change cannot be made.

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/hyperswitch_domain_models/src/payment_method_data.rs (hunk 4 above)
Comment (AkshayaFoiger) [reply crates/hyperswitch_domain_models/src/payment_method_data.rs:299]: Through stripe compatibility layer, backwards compatibility is handled.

File: crates/router/src/compatibility/stripe/payment_intents/types.rs [hunk 5]
```diff
@@ -315,6 +315,18 @@ impl TryFrom<StripePaymentIntentRequest> for payments::PaymentsRequest {
 
         let amount = item.amount.map(|amount| MinorUnit::new(amount).into());
 
+        let payment_method_data = item.payment_method_data.as_ref().map(|pmd| {
+            let payment_method_data = match pmd.payment_method_details.as_ref() {
+                Some(spmd) => Some(payments::PaymentMethodData::from(spmd.to_owned())),
+                None => get_pmd_based_on_payment_method_type(item.payment_method_types),
```
Comment (Narayanbhat166) [inline crates/router/src/compatibility/stripe/payment_intents/types.rs:321]: if no payment method data is passed we always infer `upi`?

**Review by Narayanbhat166 — APPROVED** [review]

**Review by sahkal — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Pull request merged**